protocol/consistencychecks.go

```diff
@@ -0,0 +1,328 @@
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameNotFound && proofType == proofOfAbsence:
+	case msg.Error == Success:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// determine which kind of TB's verification we should do
+	// based on the response code and proof type
+	if msg.Error == Success && proofType == proofOfAbsence {
+		// 4.1 verify returned promise
+		if err := cc.verifyReturnedPromise(requestType,
+			msg.Error, df, uname, key); err != nil {
+			return nil, err
+		}
+	} else {
+		// 4.2 verify fulfilled promise
+		if err := cc.verifyFulfilledPromise(uname, str, ap); err != nil {
+			return nil, err
+		}
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
+
+func verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+// verifySTR checks the consistency between 2 snapshots.
+// This uses the pinning signing key in cc
+// to verify the STR's signature and should not verify
+// the hash chain using the current STR stored in cc.
+func (cc *ConsistencyChecks) verifySTR(savedSTR, str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	// TODO: clearly verify which is the client's actual expectation.
+	// See #81
+	if (str.Epoch == savedSTR.Epoch && bytes.Equal(savedSTR.Signature, str.Signature)) ||
+		(str.Epoch == savedSTR.Epoch+1 && str.VerifyHashChain(savedSTR.Signature)) {
+		return nil
+	}
+
+	// TODO: verify the directory's policies as well. See #115
+	return ErrorBadSTR
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	if cc.SavedSTR.Epoch+1 != str.Epoch {
```

Comment:
Yeah, I'm not sure how this would work for all cases in which clients could miss epochs. I think for now, we're not considering these cases though since we're not supporting prior epoch lookups and monitoring for missed epochs. But we do need to rethink this verification for those operations; my take is that we should leave this as a task for the next release.